src/cpp/core/FileLockTests.cpp

```diff
@@ -0,0 +1,72 @@
+/*
+ * FileLocktests.cpp
+ *
+ * Copyright (C) 2009-14 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+#include <core/FileLock.hpp>
+
+#include <core/Error.hpp>
+#include <core/FilePath.hpp>
+
+#include <boost/thread.hpp>
+
+#define RSTUDIO_NO_TESTTHAT_ALIASES
+#include <tests/TestThat.hpp>
+
+namespace rstudio {
+namespace core {
+namespace tests {
+
+TEST_CASE("File Locking")
+{
+   SECTION("A lock can only be acquired once")
+   {
+      Error error;
+      FilePath lockFilePath("/tmp/rstudio-tests/lock");
+      
+      FileLock lock1;
+      FileLock lock2;
+      
+      error = lock1.acquire(lockFilePath);
+      if (error)
+         LOG_ERROR(error);
+      
+      CHECK((error == Success()));
+      
+      error = lock2.acquire(lockFilePath);
+      if (error)
+         LOG_ERROR(error);
+      
+      CHECK_FALSE((error == Success()));
```

Comment:
Is it possible this is because you're asking for the lock from the same process and thread that already *has* the lock? 

Does the locktester utility also fail for you?